django/db/backends/__init__.py

```diff
@@ -823,6 +823,14 @@ def cache_key_culling_sql(self):
         """
         return "SELECT cache_key FROM %s ORDER BY cache_key LIMIT 1 OFFSET %%s"
 
+    def case_cast_sql(self, output_field):
```

Comment:
I've tested the failing cases removing just the case expression to see if this isn't an interaction of fields with Value objects, or some other part of the code, but they don't fail. I've also looked at the types seen by the database. The fields start as ``unknown`` and the unified type is [resolved to ``text``](http://www.postgresql.org/docs/9.4/static/typeconv-union-case.html) as documented.

I've changed the method to accept the field instance, I can change the name to something like ``unification_cast_sql`` if that makes more sense.